ansible_collection/hpe/nimble/plugins/modules/hpe_nimble_volume.py

```diff
@@ -0,0 +1,895 @@
+
+
+def create_volume(
+        client_obj,
+        vol_name,
+        **kwargs):
+
+    if utils.is_null_or_empty(vol_name):
+        return (False, False, "Volume creation failed as volume name is null.", {})
+
+    try:
+        vol_resp = client_obj.volumes.get(id=None, name=vol_name)
+        # remove unchanged and null arguments from kwargs
+        params = utils.remove_null_args(**kwargs)
+        if utils.is_null_or_empty(vol_resp):
+            client_obj.volumes.create(vol_name, **params)
+            return (True, True, "Created volume %s successfully." % vol_name, {})
+        else:
+            return (False, False, "Volume '%s' cannot be created as it is already present." % vol_name, {})
+    except Exception as e:
+        return (False, False, "Volume creation failed | %s" % e, {})
+
+
+def delete_volume(client_obj, vol_name):
+    if utils.is_null_or_empty(vol_name):
+        return (False, False, "Volume deletion failed as volume name is null.", {})
+
+    try:
+        vol_resp = client_obj.volumes.get(id=None, name=vol_name)
+        if utils.is_null_or_empty(vol_resp):
+            return (False, False, "Volume '%s' not present to delete." % vol_name, {})
+        else:
+            # disassociate the volume from vol coll
+            client_obj.volumes.update(id=vol_resp.attrs.get("id"), volcoll_id="")
+            client_obj.volumes.offline(id=vol_resp.attrs.get("id"))
+            client_obj.volumes.delete(id=vol_resp.attrs.get("id"))
+            return (True, True, "Deleted volume %s successfully." % vol_name, {})
+    except Exception as e:
+        return (False, False, "Volume deletion failed | %s" % e, {})
+
+
+def restore_volume(client_obj, vol_name, snapshot_to_restore=None):
+    if utils.is_null_or_empty(vol_name):
+        return (False, False, "Volume restore failed as volume name is null.", {})
+    try:
+        vol_resp = client_obj.volumes.get(id=None, name=vol_name)
+        if utils.is_null_or_empty(vol_resp):
+            return (False, False, "Volume '%s' not present to restore." % vol_name, {})
+
+        if utils.is_null_or_empty(snapshot_to_restore):
+            # restore the volume to the last snapshot
+            snap_list_resp = client_obj.snapshots.list(vol_name=vol_name)
+            if utils.is_null_or_empty(snap_list_resp):
+                return (False, False, "Volume '%s' cannot be restored as no snapshot is present in source volume."
+                        % vol_name, {})
+            snap_resp = snap_list_resp[-1]
+            snapshot_to_restore = snap_resp.attrs.get("name")
+        else:
+            # get the snapshot detail from the given source vol
+            snap_resp = client_obj.snapshots.get(vol_name=vol_name, name=snapshot_to_restore)
+            if utils.is_null_or_empty(snap_resp):
+                return (False, False, "Volume '%s' cannot not be restored as given snapshot name '%s' is not present in source volume."
```

Comment:
nit: `f-strings`

`return (False, False, f"Volume '{vol_name}' cannot be restored as given snapshot name '{snapshot_to_restore}' is not present in source volume."`